### DISCUSSION TITLE
EEPROM.write and STM32F103C8T6 Blue Pill not working

### DISCUSSION THREAD

**Discussion by warp929** [stm32duino/Arduino_Core_STM32#1731] [Q&A]
I'm new to the STM32 but have some experience with emulated EEPROM on the ESP32.  After half a day of Googling and testing, I have not been able to get any of the included EEPROM examples to work on a Blue Pill (STM32F103C8T6 ).  That includes, the WRITE, READ, PUT, GET, and buffered page examples.  (I understand Read and Get won't work until writing works first.)

I'm using Arduino IDE 1.8.13 and an ST-LINK V2 for programming.  I can get other examples, like BLINK, and RC-SWITCH Advanced Demo to function properly so I know it programs. I've also erased the chip with STM32CubeProgrammer.  Please point me in the direction of what I may have missed. Thank you.

I've tried many different settings but these are the latest:
![image](https://user-images.githubusercontent.com/50374714/170934928-9a8439fd-30fb-4219-a223-29399fbc1026.png)

**Comment by fpistm** [comment]
Did you test default EEPROM examples `eeprom_put` and `eeprom_get` ?
Is it a legacy ST MCU ?

**Comment by ABOSTM** [comment]
You can check this discussion:
https://github.com/stm32duino/Arduino_Core_STM32/issues/1678

**Comment by warp929** [comment]
fpistm - Yes, I had tried the default examples.

ABOSTM - Thank you. It seems I have the same issue.  If I select STM32F103C6 for the **32K** version, then the EEPROM does function properly.

In case somebody else sees this, here is what my chip looks like.
![20220530_074230 - Copy](https://user-images.githubusercontent.com/50374714/171016587-863926e3-6926-4646-a6a4-9a72c3317288.jpg)

![Screenshot 2022-05-30 074831](https://user-images.githubusercontent.com/50374714/171016725-545d50de-47f1-4fe3-b51d-58c18d20c4a3.jpg)

**Comment by warp929** [comment]
After more research the problem is with the "stm32_eeprom.c" file.  It incorrectly calculates the Flash End at 128K for a 64K chip. Which leads to an incorrect Flash Base.  See discussion:
https://github.com/stm32duino/Arduino_Core_STM32/issues/1500#ref-commit-5c11191

The solution was to replace
C:\Users\????\AppData\Local\Arduino15\packages\STMicroelectronics\hardware\stm32\2.2.0\libraries\EEPROM\src\utility\stm32_eeprom.c
with this one:
https://github.com/fpistm/Arduino_Core_STM32/commit/5c111915f9a74f7d53c34f82135ed4dcc7fb759a

**Comment by SeongJongKwak** [comment]
@warp929

By any chance, did you solve it?
I am using 'STM32F103C6'.
If 'EEPROM' function is used, 'MCU' stops.
No further progress.

If so, how should I solve it?

**Reply by warp929** [reply]
Yes, the problem was solved for the STM32F103C8T6 and STM32G.....  
See my previous post for replacing the troublesome file.

**Reply by SeongJongKwak** [reply]
I'm sorry.
I'm curious, but I don't understand what changed. 
Could you please write in detail?
Currently I have:
https://www.stm32duino.com/viewtopic.php?t=2015

**Comment by ariefadha** [comment]
is it fixed already? my eeprom gone if i program near 100% of flash. currently in 97ish%

**Reply by fpistm** [reply]
Eeprom is emulated in flash. It uses the last flash page/sector. So if you fill all flash it could not work...

**Comment by ariefadha** [comment]
is the eeprom variable reserves on setup, so the IDE knows exactly my eeprom needs and tell that on build/compile?  if not i wonder what happened if i overflowing the eeprom variable

**Reply by fpistm** [reply]
As stated EEPROM library is not to manage a real eeprom, it emulates it in the last page of the flash.
For STM32F103C8 (bluepill) the flash size is 64kB and flash page size is 0x400 (1kB). So the EEPROM library used 1Kb.
This means that, if your application used the EEPROM library, your application can used around 63kB.
Which is your case that's all.
There is no check about the spcce used by application and the EEPROM page used.

**Reply by ariefadha** [reply]
yes i know the EEPROM is emulated from FLAHS.  what i mean is : 
the compiler shoud give warn to programmers if they have used EEPROM more than that. and or if program using FLASH that will overtake the EEPROM area.

because in program we declare type of variable taht use for EEPROM.

for example :
struct DataEE{
uint8_t foo1;
uint8_t foo2:
}

DataEE defaultVar{
50,
50
};

EEPROM.put(0, defaultVar);

so in theory, in 1Kb i can put up to1024 uint8_t foo1024. if i exceeded that the compiler should warn me. or if my program exceeded that.

what i concerned is why compiler not give any warn like when we use dedicated EEPROM in AVR chips

**Reply by fpistm** [reply]
Already answered, no warning. It is not easy to get the EEPROM size at  compile time due to the number of series supported and possible configuration. Feel free top contribute if you want to add a warning but it have to support all series.
And you issue is not you overflow the EEPROM size but your application used the last page size which is also used by the EEPROM emulation. So when the EEPROM emulation write a value, you simply overwrite a part of your application.

**Comment by ariefadha** [comment]
ok thanks for the answers
